File: src/com/android/settings/network/telephony/RoamingPreferenceController.java
/*
 * Copyright (C) 2018 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*
 * Changes from Qualcomm Innovation Center, Inc. are provided under the following license:
 * Copyright (c) 2022-2024 Qualcomm Innovation Center, Inc. All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause-Clear
 */

package com.android.settings.network.telephony;

import static androidx.lifecycle.Lifecycle.Event.ON_START;
import static androidx.lifecycle.Lifecycle.Event.ON_STOP;

import static android.telephony.ims.feature.ImsFeature.FEATURE_MMTEL;
import static android.telephony.ims.stub.ImsRegistrationImplBase.REGISTRATION_TECH_CROSS_SIM;

import android.content.Context;
import android.os.PersistableBundle;
import android.os.RemoteException;
import android.telephony.CarrierConfigManager;
import android.telephony.ims.aidl.IImsRegistration;
import android.telephony.SubscriptionInfo;
import android.telephony.SubscriptionManager;
import android.telephony.TelephonyManager;
import android.util.Log;
import android.util.SparseBooleanArray;

import androidx.annotation.VisibleForTesting;
import androidx.fragment.app.FragmentManager;
import androidx.lifecycle.LifecycleObserver;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.OnLifecycleEvent;
import androidx.preference.Preference;
import androidx.preference.PreferenceScreen;

import com.android.settings.R;
import com.android.settings.network.MobileNetworkRepository;
import com.android.settingslib.RestrictedSwitchPreference;
import com.android.settingslib.core.lifecycle.Lifecycle;
import com.android.settingslib.mobile.dataservice.MobileNetworkInfoEntity;

import java.util.ArrayList;
import java.util.List;

/**
 * Preference controller for "Roaming"
 */
public class RoamingPreferenceController extends TelephonyTogglePreferenceController implements
        LifecycleObserver, MobileNetworkRepository.MobileNetworkCallback {
    private static final String TAG = "RoamingController";
    private static final String DIALOG_TAG = "MobileDataDialog";

    private RestrictedSwitchPreference mSwitchPreference;
    private TelephonyManager mTelephonyManager;
    public SubscriptionManager mSubscriptionManager;
    private CarrierConfigManager mCarrierConfigManager;
    protected MobileNetworkRepository mMobileNetworkRepository;
    protected LifecycleOwner mLifecycleOwner;
    private List<MobileNetworkInfoEntity> mMobileNetworkInfoEntityList = new ArrayList<>();

    private DdsDataOptionStateTuner mDdsDataOptionStateTuner;
    private SparseBooleanArray mIsSubInCall;
    private SparseBooleanArray mIsCiwlanModeSupported;
    private SparseBooleanArray mIsCiwlanEnabled;
    private SparseBooleanArray mIsInCiwlanOnlyMode;
    private SparseBooleanArray mIsImsRegisteredOnCiwlan;

    @VisibleForTesting
    FragmentManager mFragmentManager;
    MobileNetworkInfoEntity mMobileNetworkInfoEntity;
    int mDialogType;

    public RoamingPreferenceController(Context context, String key, Lifecycle lifecycle,
            LifecycleOwner lifecycleOwner, int subId) {
        super(context, key);
        mSubId = subId;
        mCarrierConfigManager = context.getSystemService(CarrierConfigManager.class);
        mMobileNetworkRepository = MobileNetworkRepository.getInstance(context);
        mLifecycleOwner = lifecycleOwner;
        if (lifecycle != null) {
            lifecycle.addObserver(this);
        }
    }

    @Override
    public int getAvailabilityStatus() {
        final PersistableBundle carrierConfig = mCarrierConfigManager.getConfigForSubId(mSubId);
        if (carrierConfig != null && carrierConfig.getBoolean(
                CarrierConfigManager.KEY_FORCE_HOME_NETWORK_BOOL)) {
            return CONDITIONALLY_UNAVAILABLE;
        }
        return AVAILABLE;
    }

    @OnLifecycleEvent(ON_START)
    public void onStart() {
        mMobileNetworkRepository.addRegister(mLifecycleOwner, this, mSubId);
        mMobileNetworkRepository.updateEntity();

        // If the current instance is for the DDS, listen to the call state changes on nDDS.
        if (mSubId != SubscriptionManager.INVALID_SUBSCRIPTION_ID) {
            mDdsDataOptionStateTuner.register(mContext, mSubId);
        }
    }

    @OnLifecycleEvent(ON_STOP)
    public void onStop() {
        mMobileNetworkRepository.removeRegister(this);
        if (mSubId != SubscriptionManager.INVALID_SUBSCRIPTION_ID) {
            mDdsDataOptionStateTuner.unregister(mContext);
        }
    }

    @Override
    public void displayPreference(PreferenceScreen screen) {
        super.displayPreference(screen);
        mSwitchPreference = screen.findPreference(getPreferenceKey());
    }

    @Override
    public int getAvailabilityStatus(int subId) {
        return mSubId != SubscriptionManager.INVALID_SUBSCRIPTION_ID
                ? AVAILABLE
                : AVAILABLE_UNSEARCHABLE;
    }

    @Override
    public boolean setChecked(boolean isChecked) {
        if (isDialogNeeded()) {
            showDialog(mDialogType);
        } else {
            // Update data directly if we don't need dialog
            if (mTelephonyManager != null) {
                mTelephonyManager.setDataRoamingEnabled(isChecked);
                return true;
            }
        }

        return false;
    }

    @Override
    public void updateState(Preference preference) {
        if (mTelephonyManager == null) {
            return;
        }
        super.updateState(preference);
        mSwitchPreference = (RestrictedSwitchPreference) preference;
        update();
    }

    private void update() {
        if (mSwitchPreference == null) {
            return;
        }
        if (!mSwitchPreference.isDisabledByAdmin()) {
            mSwitchPreference.setEnabled(mSubId != SubscriptionManager.INVALID_SUBSCRIPTION_ID);
            mSwitchPreference.setChecked(isChecked());

            if (mDdsDataOptionStateTuner.isDisallowed()) {
                Log.d(TAG, "nDDS voice call in ongoing");
                // we will get inside this block only when the current instance is for the DDS
                if (isChecked()) {
                    Log.d(TAG, "Do not allow the user to turn off DDS data roaming");
                    mSwitchPreference.setEnabled(false);
                    mSwitchPreference.setSummary(
                            R.string.mobile_data_settings_summary_dds_roaming_unavailable);
                }
            }
        }
    }

    @VisibleForTesting
    boolean isDialogNeeded() {
        if (mTelephonyManager == null) {
            return false;
        }
        final boolean isRoamingEnabled = isChecked();
        final PersistableBundle carrierConfig = mCarrierConfigManager.getConfigForSubId(
                mSubId);
        // Need dialog if we need to turn on roaming and the roaming charge indication is allowed
        if (!isRoamingEnabled && (carrierConfig == null || !carrierConfig.getBoolean(
                CarrierConfigManager.KEY_DISABLE_CHARGE_INDICATION_BOOL))) {
            mDialogType = RoamingDialogFragment.TYPE_ENABLE_DIALOG;
            return true;
        }
        if (isRoamingEnabled) {
            final boolean isRoaming = MobileNetworkSettings.isRoaming(mSubId);
            final int DDS = SubscriptionManager.getDefaultDataSubscriptionId();
            final int nDDS = MobileNetworkSettings.getNonDefaultDataSub();

            // Store the call state and C_IWLAN-related settings of all active subscriptions
            int[] activeSubIdList = mSubscriptionManager.getActiveSubscriptionIdList();
            mIsSubInCall = new SparseBooleanArray(activeSubIdList.length);
            mIsCiwlanModeSupported = new SparseBooleanArray(activeSubIdList.length);
            mIsCiwlanEnabled = new SparseBooleanArray(activeSubIdList.length);
            mIsInCiwlanOnlyMode = new SparseBooleanArray(activeSubIdList.length);
            mIsImsRegisteredOnCiwlan = new SparseBooleanArray(activeSubIdList.length);
            for (int i = 0; i < activeSubIdList.length; i++) {
                int subId = activeSubIdList[i];
                TelephonyManager tm = mTelephonyManager.createForSubscriptionId(subId);
                mIsSubInCall.put(subId, tm.getCallStateForSubscription() !=
                        TelephonyManager.CALL_STATE_IDLE);
                mIsCiwlanModeSupported.put(subId, MobileNetworkSettings.isCiwlanModeSupported(
                        subId));
                mIsCiwlanEnabled.put(subId, MobileNetworkSettings.isCiwlanEnabled(subId));
                mIsInCiwlanOnlyMode.put(subId, MobileNetworkSettings.isInCiwlanOnlyMode(subId));
                mIsImsRegisteredOnCiwlan.put(subId, MobileNetworkSettings.isImsRegisteredOnCiwlan(
                        subId));
            }

            // For targets that support MSIM C_IWLAN, the warning is to be shown only for the DDS
            // when either sub is in a call. For other targets, it will be shown only when there is
            // a call on the DDS.
            boolean isMsimCiwlanSupported = MobileNetworkSettings.isMsimCiwlanSupported();
            int subToCheck = DDS;
            if (isMsimCiwlanSupported) {
                if (mSubId != DDS) {
                    // If the code comes here, the user is trying to change the roaming toggle state
                    // of the nDDS which we don't care about.
                    return false;
                } else {
                    // Otherwise, the user is trying to toggle the roaming toggle state of the DDS.
                    // In this case, we need to check if the nDDS is in a call. If it is, we will
                    // check the C_IWLAN related settings belonging to the nDDS. Otherwise, we will
                    // check those of the DDS.
                    subToCheck = subToCheckForCiwlanWarningDialog(nDDS, DDS);
                    Log.d(TAG, "isDialogNeeded DDS = " + DDS + ", subToCheck = " + subToCheck);
                }
            }

            if (isRoaming && mIsSubInCall.get(subToCheck)) {
                boolean isCiwlanModeSupported = mIsCiwlanModeSupported.get(subToCheck);
                boolean isCiwlanEnabled = mIsCiwlanEnabled.get(subToCheck);
                boolean isInCiwlanOnlyMode = mIsInCiwlanOnlyMode.get(subToCheck);
                boolean isImsRegisteredOnCiwlan = mIsImsRegisteredOnCiwlan.get(subToCheck);
                if (isCiwlanEnabled && (isInCiwlanOnlyMode || !isCiwlanModeSupported)) {
                    Log.d(TAG, "isDialogNeeded: isRoaming = true, isInCall = true, isCiwlanEnabled = true" +
                            ", isInCiwlanOnlyMode = " + isInCiwlanOnlyMode +
                            ", isCiwlanModeSupported = " + isCiwlanModeSupported +
                            ", isImsRegisteredOnCiwlan = " + isImsRegisteredOnCiwlan);
                    // If IMS is registered over C_IWLAN-only mode, the device is in a call, and
                    // user is trying to disable roaming while UE is romaing, display a warning
                    // dialog that disabling roaming will cause a call drop.
                    if (isImsRegisteredOnCiwlan) {
                        mDialogType = RoamingDialogFragment.TYPE_DISABLE_CIWLAN_DIALOG;
                        return true;
                    }
                } else {
                    Log.d(TAG, "isDialogNeeded: C_IWLAN not enabled or not in C_IWLAN-only mode");
                }
            } else {
                Log.d(TAG, "isDialogNeeded: Not roaming or not in a call");
            }
        }
        return false;
    }

    private int subToCheckForCiwlanWarningDialog(int ndds, int dds) {
        int subToCheck = SubscriptionManager.INVALID_SUBSCRIPTION_ID;
        if (mIsSubInCall.get(ndds) && mIsCiwlanEnabled.get(ndds) &&
                (mIsInCiwlanOnlyMode.get(ndds) || !mIsCiwlanModeSupported.get(ndds)) &&
                mIsImsRegisteredOnCiwlan.get(ndds)) {
            subToCheck = ndds;
        } else {
            subToCheck = dds;
        }
        return subToCheck;
    }

    @Override
    public boolean isChecked() {
        if (mTelephonyManager == null) {
            return false;
        }
        return mMobileNetworkInfoEntity == null ? false
                : mMobileNetworkInfoEntity.isDataRoamingEnabled;
    }

    public void init(FragmentManager fragmentManager, int subId, MobileNetworkInfoEntity entity) {
        mFragmentManager = fragmentManager;
        mSubId = subId;
        mMobileNetworkInfoEntity = entity;
        mTelephonyManager = mContext.getSystemService(TelephonyManager.class);
        mSubscriptionManager = mContext.getSystemService(SubscriptionManager.class);
        if (mSubId == SubscriptionManager.INVALID_SUBSCRIPTION_ID) {
            return;
        }
        final TelephonyManager telephonyManager = mTelephonyManager
                .createForSubscriptionId(mSubId);
        if (telephonyManager == null) {
            Log.w(TAG, "fail to init in sub" + mSubId);
            mSubId = SubscriptionManager.INVALID_SUBSCRIPTION_ID;
            return;
        }
        mTelephonyManager = telephonyManager;

        mDdsDataOptionStateTuner =
                new DdsDataOptionStateTuner(mTelephonyManager,
                        mSubscriptionManager,
                        () -> updateState(mSwitchPreference));
    }

    private void showDialog(int type) {
        final RoamingDialogFragment dialogFragment = RoamingDialogFragment.newInstance(
                mSwitchPreference.getTitle().toString(), type, mSubId,
                MobileNetworkSettings.isCiwlanModeSupported(mSubId));

        dialogFragment.show(mFragmentManager, DIALOG_TAG);
    }

    @VisibleForTesting
    public void setMobileNetworkInfoEntity(MobileNetworkInfoEntity mobileNetworkInfoEntity) {
        mMobileNetworkInfoEntity = mobileNetworkInfoEntity;
    }

    @Override
    public void onAllMobileNetworkInfoChanged(
            List<MobileNetworkInfoEntity> mobileNetworkInfoEntityList) {
        mMobileNetworkInfoEntityList = mobileNetworkInfoEntityList;
        mMobileNetworkInfoEntityList.forEach(entity -> {
            if (Integer.parseInt(entity.subId) == mSubId) {
                mMobileNetworkInfoEntity = entity;
                update();
                refreshSummary(mSwitchPreference);
                return;
            }
        });
    }

    @Override
    public void onDataRoamingChanged(int subId, boolean enabled) {
        if (subId != mSubId) {
            Log.d(TAG, "onDataRoamingChanged - wrong subId : " + subId + " / " + enabled);
            return;
        }
        update();
    }
}

File: src/com/android/settings/network/telephony/RoamingPreferenceController.kt
/*
 * Copyright (C) 2024 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// QTI_BEGIN: 2024-10-24: Telephony: Use kotlin implementation for roaming preference
/*
 * Changes from Qualcomm Innovation Center are provided under the following license:
 *
// QTI_END: 2024-10-24: Telephony: Use kotlin implementation for roaming preference
// QTI_BEGIN: 2025-02-12: Telephony: Fix for updating roaming data option greyed out state
 * Copyright (c) 2024-2025 Qualcomm Innovation Center, Inc. All rights reserved.
// QTI_END: 2025-02-12: Telephony: Fix for updating roaming data option greyed out state
// QTI_BEGIN: 2024-10-24: Telephony: Use kotlin implementation for roaming preference
 * SPDX-License-Identifier: BSD-3-Clause-Clear
 */
// QTI_END: 2024-10-24: Telephony: Use kotlin implementation for roaming preference
package com.android.settings.network.telephony

import android.content.Context
import android.os.UserManager
import android.util.Log
import android.telephony.CarrierConfigManager
import android.telephony.SubscriptionManager
import android.telephony.TelephonyManager
import androidx.annotation.VisibleForTesting
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.res.stringResource
import androidx.fragment.app.FragmentManager
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.android.settings.R
import com.android.settings.network.telephony.MobileNetworkSettingsSearchIndex.MobileNetworkSettingsSearchResult
import com.android.settings.network.telephony.MobileNetworkSettingsSearchIndex.MobileNetworkSettingsSearchItem
import com.android.settings.spa.preference.ComposePreferenceController
import com.android.settingslib.spa.widget.preference.SwitchPreferenceModel
import com.android.settingslib.spaprivileged.model.enterprise.Restrictions
import com.android.settingslib.spaprivileged.template.preference.RestrictedSwitchPreference

/** Preference controller for "Roaming" */
// QTI_BEGIN: 2024-10-24: Telephony: Use kotlin implementation for roaming preference
class RoamingPreferenceController
// QTI_END: 2024-10-24: Telephony: Use kotlin implementation for roaming preference
@JvmOverloads
constructor(
    context: Context,
    key: String,
    private val mobileDataRepository: MobileDataRepository = MobileDataRepository(context),
// QTI_BEGIN: 2025-02-12: Telephony: Fix for updating roaming data option greyed out state
    private val roamingPreferenceRepository: RoamingPreferenceRepository =
            RoamingPreferenceRepository(context),
// QTI_END: 2025-02-12: Telephony: Fix for updating roaming data option greyed out state
) : ComposePreferenceController(context, key) {
    @VisibleForTesting var fragmentManager: FragmentManager? = null
    private var subId = SubscriptionManager.INVALID_SUBSCRIPTION_ID

    private var telephonyManager = context.getSystemService(TelephonyManager::class.java)!!
    private val carrierConfigRepository = CarrierConfigRepository(context)
    private val roamingSearchItem = RoamingSearchItem(context)
// QTI_BEGIN: 2024-10-24: Telephony: Use kotlin implementation for roaming preference
    private var dialogType = -1
// QTI_END: 2024-10-24: Telephony: Use kotlin implementation for roaming preference

    fun init(fragmentManager: FragmentManager, subId: Int) {
        this.fragmentManager = fragmentManager
        this.subId = subId
// QTI_BEGIN: 2024-10-24: Telephony: Use kotlin implementation for roaming preference
        Log.d(TAG, "init() subId: $subId");
// QTI_END: 2024-10-24: Telephony: Use kotlin implementation for roaming preference
        telephonyManager = telephonyManager.createForSubscriptionId(subId)
// QTI_BEGIN: 2024-10-24: Telephony: Use kotlin implementation for roaming preference
        if ((this.subId == SubscriptionManager.INVALID_SUBSCRIPTION_ID) ||
                (telephonyManager == null)) {
            return;
        }
        RoamingPreferenceControllerUtil.init(mContext)
// QTI_END: 2024-10-24: Telephony: Use kotlin implementation for roaming preference
    }

    override fun getAvailabilityStatus() =
        if (roamingSearchItem.isAvailable(subId)) AVAILABLE else CONDITIONALLY_UNAVAILABLE

    @Composable
    override fun Content() {
        val summary = stringResource(R.string.roaming_enable)
        val isDataRoamingEnabled by
            remember { mobileDataRepository.isDataRoamingEnabledFlow(subId) }
                .collectAsStateWithLifecycle(null)
// QTI_BEGIN: 2025-02-12: Telephony: Fix for updating roaming data option greyed out state
        val isDisallowed by
            remember { roamingPreferenceRepository.isDisallowedFlow(subId) }
                .collectAsStateWithLifecycle(initialValue = false)
// QTI_END: 2025-02-12: Telephony: Fix for updating roaming data option greyed out state
        RestrictedSwitchPreference(
            model =
                object : SwitchPreferenceModel {
                    override val title = stringResource(R.string.roaming)
                    override val summary = { summary }
// QTI_BEGIN: 2025-02-12: Telephony: Fix for updating roaming data option greyed out state
                    override val changeable = { !isDisallowed }
// QTI_END: 2025-02-12: Telephony: Fix for updating roaming data option greyed out state
                    override val checked = { isDataRoamingEnabled }
                    override val onCheckedChange: (Boolean) -> Unit = { newChecked ->
                        if (newChecked && isDialogNeeded()) {
// QTI_BEGIN: 2024-10-24: Telephony: Use kotlin implementation for roaming preference
                            dialogType = RoamingDialogFragment.TYPE_ENABLE_DIALOG
                            showDialog(dialogType, title)
// QTI_END: 2024-10-24: Telephony: Use kotlin implementation for roaming preference
                        } else {
// QTI_BEGIN: 2024-10-24: Telephony: Use kotlin implementation for roaming preference
                            if (RoamingPreferenceControllerUtil.isDialogNeeded(subId)) {
                                dialogType = RoamingDialogFragment.TYPE_DISABLE_CIWLAN_DIALOG
                                showDialog(dialogType, title)
                            } else {
                                // Update data directly if we don't need dialog
                                telephonyManager.isDataRoamingEnabled = newChecked
                            }
// QTI_END: 2024-10-24: Telephony: Use kotlin implementation for roaming preference
                        }
                    }
                },
            restrictions = Restrictions(keys = listOf(UserManager.DISALLOW_DATA_ROAMING)),
        )
    }

    @VisibleForTesting
    fun isDialogNeeded(): Boolean {
// QTI_BEGIN: 2024-10-24: Telephony: Use kotlin implementation for roaming preference
        if (telephonyManager == null) {
            return false;
        }

// QTI_END: 2024-10-24: Telephony: Use kotlin implementation for roaming preference
        // Need dialog if we need to turn on roaming and the roaming charge indication is allowed
        return !carrierConfigRepository.getBoolean(
            subId, CarrierConfigManager.KEY_DISABLE_CHARGE_INDICATION_BOOL)
    }

// QTI_BEGIN: 2024-10-24: Telephony: Use kotlin implementation for roaming preference
    private fun showDialog(type: Int, preftitle: String) {
        Log.d(TAG, "showDialog type: $type")
        fragmentManager?.let { RoamingDialogFragment.newInstance(preftitle, type, subId,
                MobileNetworkSettings.isCiwlanModeSupported(subId)).show(it, DIALOG_TAG) }
// QTI_END: 2024-10-24: Telephony: Use kotlin implementation for roaming preference
    }

    companion object {
        private const val DIALOG_TAG = "MobileDataDialog"
// QTI_BEGIN: 2024-10-24: Telephony: Use kotlin implementation for roaming preference
        private const val TAG = "RoamingPreferenceController"
// QTI_END: 2024-10-24: Telephony: Use kotlin implementation for roaming preference

        class RoamingSearchItem(private val context: Context) : MobileNetworkSettingsSearchItem {
            private val carrierConfigRepository = CarrierConfigRepository(context)

            fun isAvailable(subId: Int): Boolean =
                SubscriptionManager.isValidSubscriptionId(subId) &&
                    !carrierConfigRepository.getBoolean(
                        subId, CarrierConfigManager.KEY_FORCE_HOME_NETWORK_BOOL)

            override fun getSearchResult(subId: Int): MobileNetworkSettingsSearchResult? {
                if (!isAvailable(subId)) return null
                return MobileNetworkSettingsSearchResult(
                    key = "button_roaming_key",
                    title = context.getString(R.string.roaming),
                )
            }
        }
    }
}
